File: src/cache_node/CacheNode.cc
#include "CacheNode.h"

using namespace cache_node;

// -------------------------------------------------------------------------------------------------
// Constructors and destructors

CacheNode::CacheNode(const string &node_id, bool is_server) : 
    AtomSpaceNode(node_id, LeadershipBrokerType::SINGLE_MASTER_SERVER, MessageBrokerType::GRPC) {

    this->is_server = is_server;
}

CacheNodeServer::CacheNodeServer(const string &node_id) : CacheNode(node_id, true) {
}

CacheNodeClient::CacheNodeClient(
    const string &node_id, 
    string &server_id) : CacheNode(node_id, false) {

    this->server_id = server_id;
    this->add_peer(server_id);
}

CacheNode::~CacheNode() {
}

CacheNodeServer::~CacheNodeServer() {
}

CacheNodeClient::~CacheNodeClient() {
}

// -------------------------------------------------------------------------------------------------
// Public API delegated to concrete nodes

void CacheNodeServer::node_joined_network(const string &node_id) {
    this->add_peer(node_id);
}

void CacheNodeClient::node_joined_network(const string &node_id) {
    // Do nothing
}

string CacheNodeServer::cast_leadership_vote() {
    return this->node_id();
}

string CacheNodeClient::cast_leadership_vote() {
    return this->server_id;
}

// -------------------------------------------------------------------------------------------------
// Other methods

Message *CacheNode::message_factory(string &command, vector<string> &args) {
    Message *message = AtomSpaceNode::message_factory(command, args);
    if (message != NULL) {
        return message;
    }
    return NULL;
}

File: examples/cache_node.py
from hyperon_das_node import AtomSpaceNode, Message, LeadershipBrokerType, MessageBrokerType

class PrintMessage(Message):
    def __init__(self, content: str):
        super().__init__()
        self.content = content

    def act(self, node: "CacheNode") -> None:
        # ideally we should call a node.method in here 
        node.print_content(self.content)

class CacheNode(AtomSpaceNode):
    def __init__(self, node_id: str, is_server: bool) -> None:
        super().__init__(
            node_id,
            LeadershipBrokerType.SINGLE_MASTER_SERVER,
            MessageBrokerType.GRPC,
        )

        self.is_server = is_server
        self.known_commands = {
            "print": PrintMessage,
        }

    def print_content(self, content: str):
        print(content)

    def message_factory(self, command: str, args: list[str]) -> Message:
        message = super().message_factory(command, args)
        if message is not None:
            return message

        message = PrintMessage(content=args[0])
        return message

        return None

class CacheNodeServer(CacheNode):
    def __init__(self, node_id: str) -> None:
        super().__init__(node_id, True)

    def node_joined_network(self, node_id: str) -> None:
        self.add_peer(node_id)

    def cast_leadership_vote(self) -> str:
        return self.node_id()

class CacheNodeClient(CacheNode):
    def __init__(self, node_id: str, server_id: str) -> None:
        super().__init__(node_id, False)
        self.server_id = server_id
        self.add_peer(server_id)

    def node_joined_network(self, node_id: str) -> None:
        # do nothing
        pass

    def cast_leadership_vote(self) -> str:
        return self.server_id


if __name__ == "__main__":
    server = CacheNodeServer("localhost:35700")
    client = CacheNodeClient("localhost:35701", "localhost:35700")
    server.join_network()
    client.join_network()

    client.send("print", ["something"], "localhost:35700")
